horovod/common/process_sets.py

```diff
@@ -0,0 +1,171 @@
+from typing import *
+# for type annotations, importing mpi4py has dangerous side effects
+try:
+    from horovod.common.basics import HorovodBasics
+except ImportError:
+    class HorovodBasics:
+        ...
+class MPI:
+    class Comm:
+        ...
+
+from horovod.common.util import is_iterable
+
+
+_basics = None  # type: Optional[HorovodBasics]
+
+
+class ProcessSet:
+    """ Representation of a set of Horovod processes that will run collective operations together
+
+    Initialize a ProcessSet with a list of process ranks or an MPI communicator. Then pass this instance to hvd.init()
+    or hvd.add_process_set().
+    """
+    process_set_id = None
+    ranks = None
+    mpi_comm = None
+
+    def __init__(self, ranks_or_comm: Union[Sequence[int], MPI.Comm]):
+        if is_iterable(ranks_or_comm):
+            ranks_or_comm = sorted(ranks_or_comm)
+            if any(not isinstance(rk, int) for rk in ranks_or_comm):
+                raise ValueError(
+                    "ProcessSet should be initialized with a list of process ranks or an mpi4py Comm object")
+            self.ranks = ranks_or_comm
+        else:
+            assert _basics is not None, "process_sets.setup() must be called first"
+            if not _basics.mpi_built():
+                raise ValueError(
+                    "Horovod has not been built with MPI support. Ensure MPI is installed and "
+                    "reinstall Horovod with HOROVOD_WITH_MPI=1 to debug the build error.")
+            from mpi4py import MPI
+            if not isinstance(ranks_or_comm, MPI.Comm):
+                raise ValueError(
+                    "ProcessSet should be initialized with a list of process ranks or an mpi4py Comm object")
+            self.mpi_comm = ranks_or_comm
+
+    def invalidate(self):
+        self.process_set_id = None
+        self.ranks = None
+        self.mpi_comm = None
+
+    def size(self) -> Optional[int]:
+        """ Return size of the process set or None if not initialized. """
+        if self.process_set_id is None:
+            return None
+        return _basics.process_set_size(self.process_set_id)
+
+    def included(self) -> Optional[bool]:
+        """ Return whether the current process is part of this process set or None if not initialized. """
+        if self.ranks is None:
+            return None
+        return _basics.rank() in self.ranks
+
+    def __str__(self) -> str:
+        return f"ProcessSet(process_set_id={self.process_set_id}, ranks={self.ranks}, mpi_comm={self.mpi_comm})"
+
+
+global_process_set = ProcessSet([])
+global_process_set.process_set_id = 0
+
+_id_to_process_sets: Dict[int, List[ProcessSet]] = {0: [global_process_set]}
```

Comment:
My internal resasoning was basically that if a user managed to create lots of equivalent `ProcessSet` instances, then removing one of them should either invalidate all of them (current implementation) or pop that handle of the list until it's empty and the internal process set should actually be removed. The simpler third alternative is, of course, to disallow the creation of lots of equivalent process set objects.

Reply:
Should multiple process sets containing the same set of processes be considered "equivalent" or should they be distinct? For example, if I were to run something like:
```
ps1 = hvd.add_process_set([0,1,2,3])
ps2 = hvd.add_process_set([0,1,2,3])
hvd.remove_process_set(ps1)
```
should `ps2` still work? I think it should, but it sounds like in the current implementation that isn't the case. 

Is there an issue with allowing a user to create multiple distinct process sets that contain the same set of processes (similar to create a bunch of MPI subcommunicators containing the same ranks)? I understand from a performance standpoint, you may want to back the "equivalent" process sets with the same communicator in the backend, but depending on the complexity of this code to allow for this kind of aliasing, it might be easier to just create multiple backend process sets, even if they contain the same sets of ranks. You can always print a warning if this occurs to tell users that performance may be degraded if they do this.

Reply:
> disallow the creation of lots of equivalent process set objects

I went with this and removed the global dict.

Reply:
Sorry, I missed your message in this thread.

Do you think it's really important to allow for that flexibility? I wouldn't want users to accidentally add the same backend process set again and again in their training loops. 

(As of my latest commit the second add_process_set call would now raise an exception.)

Reply:
Nope, not important to allow for that, and disallowing it works (and is simpler). Important point was just to make sure individual process set objects (even if containing the same ranks), do not interact with each other. Since you now cannot have multiple process sets containing the same ranks, this skirts past the issue altogether. 👍

Reply:
Cool, I'm growing to appreciate simplicity more and more. I guess just having extra ids for duplicate process sets in the backend would be the best way to go after all if we ever decide to go that way. Trying to track an additional object-id correspondence in the frontend comes with nasty  surprises as you pointed out.